src/hooks/useQRCodeGenerator.jsx

```diff
@@ -1,78 +1,95 @@
+          // Updates the relevant state variables with the generated QR code data URL and adjusts the UI accordingly.
+          setQr(url);
+          // Considers the user experience and how the UI should change once the QR code has been generated.
+          setShowInput(false);
+        }
+      );
+    }
   };
-
-  // Function to download the generated QR code as a PNG file
+  // Function to download the generated QR code as a PNG file.
   const downloadQRCode = () => {
-    // HINT 1: Consider encapsulating the filename prompting logic into a separate function.
+    // Encapsulates the filename prompting logic into a separate function.
     const getFileName = () => {
-      // HINT 2: Use a method to prompt the user for input and store the response.
-      // ...
-      // HINT 3: Implement a check for an empty filename and utilize recursion to re-prompt the user if necessary.
-      // ...
-      // HINT 4: Ensure the function returns the obtained filename.
-      // ...
+      // Uses a method to prompt the user for input and store the response.
+      // Prompts the user for a filename.
+      const fileName = prompt("Please enter a name for the downloaded file");
+      // Implements a check for an empty filename and utilizes recursion to re-prompt the user if necessary.
+      // Ensures the function returns the obtained filename.
+      return fileName === "" ? getFileName() : fileName;
     };
-
-    // HINT 5: Call the above function to retrieve a filename and store it in a variable.
-    // ...
-
-    // HINT 6: Format the filename to ensure it is filesystem-friendly.
-    // ...
-
-    // HINT 7: Create an anchor element to facilitate the download.
-    // ...
-
-    // HINT 8: Set the necessary attributes on the anchor element to prepare it for download.
-    // ...
-
-    // HINT 9: Append the anchor element to the document to make it interactable.
-    // ...
-
-    // HINT 10: Programmatically trigger a click on the anchor element to initiate the download.
-    // ...
-
-    // HINT 11: Remove the anchor element from the document after the download has been initiated.
-    // ...
+    // Calls the above function to retrieve a filename and stores it in a variable.
+    let fileName = getFileName();
+    // Formats the filename to ensure it is filesystem-friendly.
+    fileName = fileName.split(" ").join("-");
+    // Creates an anchor element to facilitate the download.
+    const link = document.createElement("a");
+    // Sets the necessary attributes on the anchor element to prepare it for download.
+    link.href = qr;
+    link.download = `${fileName}.png`;
+    // Appends the anchor element to the document to make it interactable.
+    document.body.appendChild(link);
+    // Programmatically triggers a click on the anchor element to initiate the download.
+    link.click();
+    // Removes the anchor element from the document after the download has been initiated.
+    document.body.removeChild(link);
```

Comment:
Super clear comments that makes it very easy to follow and understand what's going on!